packages/flutter/lib/src/material/navigation_bar.dart

```diff
@@ -0,0 +1,1164 @@
+/// navigation bar destinations.
+enum NavigationDestinationLabelBehavior {
+  /// Always shows all of the labels under each navigation bar destination,
+  /// selected and unselected.
+  alwaysShow,
+
+  /// Never shows any of the labels under the navigation bar destinations,
+  /// regardless of selected vs unselected.
+  alwaysHide,
+
+  /// Only shows the labels of the selected navigation bar destination.
+  ///
+  /// When a destination is unselected, the label will be faded out, and the
+  /// icon will be centered.
+  ///
+  /// When a destination is selected, the label will fade in and the label and
+  /// icon will slide up so that they are both centered.
+  onlyShowSelected,
+}
+
+/// Destination Widget for displaying Icons + labels in the Material 3
+/// Navigation Bars through [NavigationBar.destinations].
+///
+/// The destination this widget creates will look something like this:
+/// =======
+/// |
+/// |  ☆  <-- [icon] (or [selectedIcon])
+/// | text <-- [label]
+/// |
+/// =======
+///
+/// See also:
+///  * [NavigationDestinationBuilder] - Use this class when you need more
+///    custom destinations than [NavigationDestination] provides, for
+///    example, if you have animated icons.
+class NavigationDestination extends StatelessWidget {
+  /// Creates a navigation bar destination with an icon and a label, to be used
+  /// in the [NavigationBar.destinations].
+  const NavigationDestination({
+    Key? key,
+    required this.icon,
+    this.selectedIcon,
+    required this.label,
+    this.tooltip,
+  }) : super(key: key);
+
+  /// The [Widget] (usually an [Icon]) that displays for this
+  /// [NavigationDestination].
+  final Widget icon;
+
+  /// The optional [Widget] (usually an [Icon]) that displays when this
+  /// [NavigationDestination] is selected.
+  ///
+  /// If [selectedIcon] is non-null, the destination will fade from
+  /// [icon] to [selectedIcon] when this destination goes from unselected to
+  /// selected.
+  final Widget? selectedIcon;
+
+  /// The text label that appears below the icon of this
+  /// [NavigationDestination].
+  final String label;
+
+  /// The text to display in the tooltip for this [NavigationDestination], when
+  /// the user long presses the destination.
+  ///
+  /// If [tooltip] is an empty string, no tooltip will be used.
+  ///
+  /// Defaults to null, in which case the [label] text will be used.
+  final String? tooltip;
+
+  @override
+  Widget build(BuildContext context) {
+    final ThemeData theme = Theme.of(context);
+    final ColorScheme colorScheme = theme.colorScheme;
+    final NavigationBarThemeData navigationBarTheme = NavigationBarTheme.of(context);
+    final Animation<double> animation =
+        NavigationDestinationInfo.of(context).selectedAnimation;
+
+    return NavigationDestinationBuilder(
+      label: label,
+      tooltip: tooltip,
+      buildIcon: (BuildContext context) {
+        final Widget selectedIconWidget = IconTheme.merge(
```

Comment:
I don't think that would work since we need 2 different icon themes depending on the selected vs unselected icons.

I think if the icon widgets have text, then it could be left up to the developer to specify text styles, since that would not be the standard usage for this widget.